xapian-letor/ranker/listmle_ranker.cc

```diff
@@ -126,12 +126,12 @@ batch_learning(vector<FeatureVector> feature_vectors,
 }
 
 void
-ListMLERanker::train(const vector<FeatureVector>& training_data)
+ListMLERanker::train(const vector<vector<FeatureVector>>& training_data)
 {
     LOGCALL_VOID(API, "ListMLERanker::train", training_data);
     if (training_data.empty())
 	throw InvalidArgumentError("Cannot train: no training data");
-    int feature_cnt = training_data[0].get_fcount();
+    int feature_cnt = training_data[0][0].get_fcount();
```

Comment:
What about the `training_data[0].empty()` case?

Reply:
> Or having different number of features than the rest of the vectors ( We should have a better naming convention here, each instance corresponds to a query ). Although in the current use that seems unlikely, I think that would be a good check in case anyone decides to use ListMLERanker::train outside Ranker::train_model.

>
I really think that is not necessary in this case but I might be wrong.

Reply:
We ought to at least ensure something sensible happens if there are different numbers of entries in some `FeatureVector` objects compared to others.  That could be throwing an exception, or we could do something like assuming that any missing features compared to the first one are zero and any extra features get ignored.  It definitely shouldn't mean reading or writing off the end of some vector or using uninitialised values in the calculations.

It's unclear to me what actually happens with the current code.

Reply:
Pretty easy to write a test for vectors of differing lengths, as well.

Reply:
fixed in a fixup commit.